### PR TITLE
More complete cancel test

### PR DESCRIPTION
**Pull request by yschimke**

### CODE REVIEW

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+        }
+      }
+    if (connectionType != HTTP) {
+      server.useHttps(handshakeCertificates.sslSocketFactory(), false)
+    }
+    server.start()
+
+    client = clientTestRule.newClientBuilder()
+        .socketFactory(object : DelegatingSocketFactory(SocketFactory.getDefault()) {
+          @Throws(
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt:116]: off topic: I need to somehow convince spotless that this is bad style

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+        }
+        .build()
+    threadToCancel = Thread.currentThread()
+  }
+
+  @Test
+  fun cancelWritingRequestBody() {
+    server.enqueue(MockResponse())
+    val call = client.newCall(
+        Builder()
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt:137]: FWIW I dislike direct imports of widely-overloaded class names like `Builder`.

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+            .build()
+    )
+    cancelLater(call, 500)
+    try {
+      call.execute()
+      fail()
+    } catch (expected: IOException) {
+      assertEquals(cancelMode == INTERRUPT, Thread.interrupted())
+    }
+    assertEquals(if (connectionType == H2) 1 else 0, client.connectionPool.connectionCount())
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt:164]: nice

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+  @Test
+  fun cancelReadingResponseBody() {
+    val responseBodySize = 8 * 1024 * 1024 // 8 MiB.
+    server.enqueue(
+        MockResponse()
+            .setBody(
+                Buffer()
+                    .write(ByteArray(responseBodySize))
+            )
+            .throttleBody(64 * 1024.toLong(), 125, MILLISECONDS)
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt:176]: 64L * 1024L

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+    val buffer = ByteArray(1024)
+    try {
+      while (responseBody.read(buffer) != -1) {
+      }
+      fail("Expected connection to be closed")
+    } catch (expected: IOException) {
+      assertEquals(cancelMode == INTERRUPT, Thread.interrupted())
+    }
+    responseBody.close()
+    assertEquals(if (connectionType == H2) 1 else 0, client.connectionPool.connectionCount())
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt:227]: This is tested above, right? Seems unnecessary

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+
+    val call2 = client.newCall(Builder().url(server.url("/")).build())
+    call2.execute().use {
+      assertEquals(".", it.body!!.string())
+    }
+
+    val events = listener.eventSequence.filter { isConnectionEvent(it) }.map { it.name }
+
+    val expectedEvents = when {
+      cancelMode == CANCEL && connectionType == H2 -> listOf("CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "Canceled", "ResponseFailed", "ConnectionReleased", "CallStart", "ConnectionAcquired", "ConnectionReleased", "CallEnd")
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt:237]: build this in steps to make it more legible? As-is it’s not obvious where the deltas are.

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+      cancelMode == CANCEL && connectionType == HTTP -> listOf("CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "Canceled", "ResponseFailed", "ConnectionReleased", "CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "ConnectionReleased", "CallEnd")
+      cancelMode == INTERRUPT && connectionType == H2 -> listOf("CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "ResponseFailed", "ConnectionReleased", "CallStart", "ConnectionAcquired", "ConnectionReleased", "CallEnd")
+      cancelMode == INTERRUPT && connectionType == HTTPS -> listOf("CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "ResponseFailed", "ConnectionReleased", "CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "ConnectionReleased", "CallEnd")
+      cancelMode == INTERRUPT && connectionType == HTTP -> listOf("CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "ResponseFailed", "ConnectionReleased", "CallStart", "ConnectStart", "ConnectEnd", "ConnectionAcquired", "ConnectionReleased", "CallEnd")
+      else -> listOf()
+    }
+
+    assertEquals(expectedEvents, events)
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt:246]: Use assertJ so we get nicer failure messages?

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+  }
+
+  private fun isConnectionEvent(it: CallEvent?) =
+    it is CallStart || it is CallEnd || it is ConnectStart || it is ConnectEnd || it is ConnectionAcquired || it is ConnectionReleased || it is Canceled || it is RequestFailed || it is ResponseFailed
```
Comment (swankjesse) [inline okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt:250]: wrap?

**Comment by swankjesse** [comment]
Only thing to add next is a response cache parameter!

File: okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt (hunk 6 above)
Comment (yschimke) [reply okhttp/src/test/java/okhttp3/internal/http/CancelTest.kt]: Good call, turned out much nicer.

**Pull request merged**